src/com/intellij/plugins/bodhi/pmd/PMDInvoker.java

```diff
@@ -68,33 +71,53 @@ public void runPMD(AnActionEvent actionEvent, String rule, boolean isCustomRuleS
 
+        Project project = actionEvent.getData(DataKeys.PROJECT);
+        PMDProjectComponent projectComponent = project.getComponent(PMDProjectComponent.class);
+
         List<File> files = new LinkedList<File>();
-        if (actionEvent.getPlace().equals(ActionPlaces.PROJECT_VIEW_POPUP)) {
+        if (actionEvent.getPlace().equals(ActionPlaces.PROJECT_VIEW_POPUP)
+                || actionEvent.getPlace().equals(ActionPlaces.SCOPE_VIEW_POPUP)
+                || actionEvent.getPlace().equals(ActionPlaces.CHANGES_VIEW_POPUP)
+                || actionEvent.getPlace().equals(ActionPlaces.MAIN_MENU)
+                ) {
+
             //If selected by right-click on file/folder (s)
-            VirtualFile[] selectedFiles = actionEvent.getData(DataKeys.VIRTUAL_FILE_ARRAY);
-            if (selectedFiles.length == 0) {
-                //toolWindow.displayErrorMessage("Please select a file to process first");
-                return;
+            VirtualFile[] selectedFiles;
+            if(actionEvent.getPlace().equals(ActionPlaces.CHANGES_VIEW_POPUP))
+            {
+                selectedFiles = VcsContextFactory.SERVICE.getInstance().createContextOn(actionEvent).getSelectedFiles();
             }
-            for (int i = 0; i < selectedFiles.length; i++) {
-                //Add all java files recursively
-                PMDUtil.listFiles(new File(selectedFiles[i].getPresentableUrl()), files,
-                        PMDUtil.createFileExtensionFilter("java"));
+            else if(actionEvent.getPlace().equals(ActionPlaces.MAIN_MENU))
+            {
+                VirtualFile[] contentRoots = ProjectRootManager.getInstance(project).getContentRoots();
+                selectedFiles = VfsUtil.getCommonAncestors(contentRoots);
             }
-        } else if (actionEvent.getPlace().equals("ChangesViewPopup")) {
-            VirtualFile[] selectedFiles = VcsContextFactory.SERVICE.getInstance().createContextOn(actionEvent).getSelectedFiles();
-            if (selectedFiles.length == 0) {
+            else
+            {
+                selectedFiles = actionEvent.getData(DataKeys.VIRTUAL_FILE_ARRAY);
+            }
+
+            if (selectedFiles == null || selectedFiles.length == 0) {
                 //toolWindow.displayErrorMessage("Please select a file to process first");
                 return;
             }
-            for (int i = 0; i < selectedFiles.length; i++) {
+            List<VirtualFileFilter> filters = new ArrayList<>();
+            filters.add(fileHasExtension("java"));
+            filters.add(fileInSources(project));
+            if(projectComponent.isSkipTestSources())
+            {
+                filters.add(VirtualFileFilters.not(fileInTestSources(project)));
+            }
+            VirtualFileFilter filter = VirtualFileFilters.or(
+                    isDirectory(),
+                    VirtualFileFilters.and(filters.toArray(new VirtualFileFilter[filters.size()]))
```

Comment:
[Minor] It will be easier to read if you pass the filters directly as arguments to `and()`. Something like `and(fileHasExtension("java"), fileInSources(project)` etc. If you pass projectcomponent to isSkipTestSources, the conditional check can be done inside the filter as well.